### ISSUE TITLE
flutter1.2.8升级到1.2.9出现数据加载失败问题

### ISSUE THREAD

**Issue by chowjee** [Tencent/MMKV#690]
### The language of MMKV

> dart

### The version of MMKV

>  v1.2.9  

### The platform of MMKV

> Android

### The installation of MMKV

> pub.dev

### What's the issue?
MMKV flutter 1.2.8版本升级到1.2.9版本出现数据加载失败问题

### What's the log of MMKV when that happened?

// flutter mmkv 1.2.8版本

2021-05-25 15:21:25.692 15698-15698/? I/MMKV: <native-bridge.cpp:95::JNI_OnLoad> current API level = 29
2021-05-25 15:21:25.693 15698-15698/? I/MMKV: <MMKV.cpp:192::initialize> page size:4096
2021-05-25 15:21:25.693 15698-15698/? I/MMKV: <MMKV.cpp:206::initializeMMKV> root dir: /data/user/0/com.tencent.mm/files/mmkv
2021-05-25 15:21:25.697 15717-15717/? I/MMKV: <native-bridge.cpp:95::JNI_OnLoad> current API level = 29
2021-05-25 15:21:25.698 15717-15717/? I/MMKV: <MMKV.cpp:192::initialize> page size:4096
2021-05-25 15:21:25.698 15717-15717/? I/MMKV: <MMKV.cpp:206::initializeMMKV> root dir: /data/user/0/com.tencent.mm/files/mmkv
2021-05-25 15:21:30.854 16386-16386/? I/MMKV: <native-bridge.cpp:95::JNI_OnLoad> current API level = 29
2021-05-25 15:21:30.855 16386-16386/? I/MMKV: <MMKV.cpp:192::initialize> page size:4096
2021-05-25 15:21:30.855 16386-16386/? I/MMKV: <MMKV.cpp:206::initializeMMKV> root dir: /data/user/0/com.tencent.mm/files/mmkv
2021-05-25 15:21:31.170 16527-16527/? I/MMKV: <native-bridge.cpp:95::JNI_OnLoad> current API level = 29
2021-05-25 15:21:31.171 16527-16527/? I/MMKV: <MMKV.cpp:192::initialize> page size:4096
2021-05-25 15:21:31.171 16527-16527/? I/MMKV: <MMKV.cpp:206::initializeMMKV> root dir: /data/user/0/com.tencent.mm/files/mmkv
2021-05-25 15:21:32.305 15761-16243/com.test I/MMKV: <MMKV.cpp:156::initialize> version v1.2.8 page size:4096
2021-05-25 15:21:32.306 15761-16243/com.test I/MMKV: <MMKV.cpp:167::initialize> armv8 AES instructions is supported
2021-05-25 15:21:32.306 15761-16243/com.test I/MMKV: <MMKV.cpp:175::initialize> armv8 CRC32 instructions is supported
2021-05-25 15:21:32.306 15761-16243/com.test I/MMKV: <MMKV.cpp:198::initializeMMKV> root dir: /data/user/0/com.test/app_flutter/mmkv
2021-05-25 15:21:32.312 15761-16289/com.test I/Flutter-Test: [, , 0]:MMKV for flutter with rootDir = /data/user/0/com.test/app_flutter/mmkv
2021-05-25 15:21:32.315 15761-16243/com.test I/MMKV: <MMKV_IO.cpp:81::loadFromFile> loading [mmkv.default] with 10263 actual size, file size 16384, InterProcess 0, meta info version:3
2021-05-25 15:21:32.315 15761-16243/com.test I/MMKV: <MMKV_IO.cpp:86::loadFromFile> loading [mmkv.default] with crc 1748640544 sequence 2 version 3
**2021-05-25 15:21:32.315 15761-16243/com.test I/MMKV: <MMKV_IO.cpp:130::loadFromFile> loaded [mmkv.default] with 21 key-values**
2021-05-25 15:21:32.327 15761-16289/com.test I/Flutter-Test: [, , 0]:MMKV all keys: [key_has_processed_legacy_shown_conversation_13483184, key_user_cahnge_shield_modal_select_13483184, key_user_shield_modal_default_close_13483184, key_current_user_location_info, key_user_shield_modal_duration_13483184, key_user_shield_modal_timestamp_13483184, key_location_permission_granted, key_First_show_advanced_filter_page_13483184_2021年05月25日, key_last_logon_user, key_last_logid, key_last_click_nearby_tab, key_risk_control_0}, key_user_data_show_toast, key_loaded_index, key_avatar_toast, key_user_avatar_show_toast, key_risk_control_13483184}, key_data_toast, key_friend_avatar_show_toast, key_show_authentication, login_index_select_protocol]

// flutter mmkv 1.2.9版本
2021-05-25 15:24:19.345 17234-17564/com.test I/MMKV: <MMKV.cpp:156::initialize> version v1.2.8 page size:4096
2021-05-25 15:24:19.345 17234-17564/com.test I/MMKV: <MMKV.cpp:167::initialize> armv8 AES instructions is supported
2021-05-25 15:24:19.345 17234-17564/com.test I/MMKV: <MMKV.cpp:175::initialize> armv8 CRC32 instructions is supported
2021-05-25 15:24:19.345 17234-17564/com.test I/MMKV: <MMKV.cpp:198::initializeMMKV> root dir: /data/user/0/com.test/app_flutter/mmkv
2021-05-25 15:24:19.348 17234-17588/com.test I/Flutter-Test: [, , 0]:MMKV for flutter with rootDir = /data/user/0/com.test/app_flutter/mmkv
2021-05-25 15:24:19.350 17234-17564/com.test I/MMKV: <MMKV_IO.cpp:81::loadFromFile> loading [mmkv.default] with 15803 actual size, file size 16384, InterProcess 0, meta info version:3
2021-05-25 15:24:19.350 17234-17564/com.test I/MMKV: <MMKV_IO.cpp:86::loadFromFile> loading [mmkv.default] with crc 3162175340 sequence 2 version 3
**2021-05-25 15:24:19.351 17234-17564/com.test E/MMKV: <MiniPBCoder.cpp:328::decodeOneMap> InvalidProtocolBuffer truncatedMessage**
**2021-05-25 15:24:19.351 17234-17564/com.test I/MMKV: <MMKV_IO.cpp:130::loadFromFile> loaded [mmkv.default] with 0 key-values**
2021-05-25 15:24:19.357 17234-17588/com.test I/Flutter-Test: [, , 0]:MMKV all keys: []

**Comment by lingol** [comment]
Can you reproduce this bug on your device? If yes, please provide the mmkv file & crc file just **before** upgrade to v1.2.9.

**Comment by chowjee** [comment]
[mmkv.zip](https://github.com/Tencent/MMKV/files/6537483/mmkv.zip)
在我的oppo手机上是必现的问题，这是升级之前的mmkv文件。

**Comment by lingol** [comment]
Did you **encrypt** the MMKV in v1.2.8, yet forget to set that ecnryption key on v1.2.9?

**Comment by lingol** [comment]
In fact, both of your logs shows that you are using **the same MMKV Flutter v1.2.9**. So I don't think it has anything todo with the upgrading.

The one that loads successfully.
>2021-05-25 15:21:32.305 15761-16243/com.test I/MMKV: MMKV.cpp:156::initialize version v1.2.8 page size:4096

The one that fails to load.
>2021-05-25 15:24:19.345 17234-17564/com.test I/MMKV: MMKV.cpp:156::initialize version v1.2.8 page size:4096

**Comment by chowjee** [comment]
> In fact, both of your logs shows that you are using **the same MMKV Flutter v1.2.9**. So I don't think it has anything todo with the upgrading of MMKV.
> 
> The one that loads successfully.
> 
> > 2021-05-25 15:21:32.305 15761-16243/com.test I/MMKV: MMKV.cpp:156::initialize version v1.2.8 page size:4096
> 
> The one that fails to load.
> 
> > 2021-05-25 15:24:19.345 17234-17564/com.test I/MMKV: MMKV.cpp:156::initialize version v1.2.8 page size:4096

没有做加密处理。
从github提及记录来看，mmkv flutter1.2.8升级到1.2.9版本并没涉及到native层的改动，只提交了dart ffi的适配，这里输出v1.2.8是不是因为这个原因呢？

**Comment by lingol** [comment]
Are you sure about that? Because when opened in a hex editor, I don't see any regular ASCII char of the keys you printed. And that strongly suggests that it's been encrypted.

I have tried to load it in Android directly, without using the MMKV for Flutter wrapper, and it still fails. So it has nothing to do with MMKV for Flutter, or dart ffi.

**Comment by lingol** [comment]
MMKV for Flutter is **one version newer** than the native lib. So v1.2.8 means MMKV for Flutter v1.2.9, and v1.2.7 means MMKV for Flutter v1.2.8, etc.

**Comment by chowjee** [comment]
> Are you sure about that? Because when opened in a hex editor, I don't see any regular ASCII char of the keys you printed. And that strongly suggests that it's been encrypted.
> 
> I have tried to load it in Android directly, without using the MMKV for Flutter wrapper, and it still fails. So it has nothing to do with MMKV for Flutter, or dart ffi.

我确定业务代码是通过**_mmkv = MMKV.defaultMMKV();** 创建并使用mmkv的，该方法默认是不加密的；
关于日志输出的版本，我也感到很疑惑，因为我是通过pub dev引用的mmkv，分别引用的**mmkv: 1.2.8**和**mmkv: 1.2.9**，但日志输出的都是v1.2.8

**Comment by lingol** [comment]
There are only two possible explanations. 
1. Either you will fail to load it at a second launch using MMKV for Flutter v1.2.8. And that means something has gone wrong before the second launch.
2. Or, you load it successfully at a second launch. And that means you must have set an encryption key somewhere and you don't yet notice it.

**Comment by chowjee** [comment]
> I have tried to load your uploaded files with a demo project using MMKV for Flutter v1.2.8. Guess what, it fails as well.
> Have you tried to load the same file again, using **the same MMKV** for Flutter v1.2.8?

mmkv文件中保存了登录access token，在没升级mmkv的版本(1.2.8)中,可以正常读取，关于这点我们的产品已经在线上运行很多个版本了，所以可以确认升级之前是正常使用的。

**Comment by lingol** [comment]
And how do you explain that it fails to load using MMKV for Flutter v1.2.8 on my side?

So it's the second explanation that you were using an encryption key?

**Comment by chowjee** [comment]
> And how do you explain that it fails to load using MMKV for Flutter v1.2.8 on my side?
> 
> So it's the second explanation that you were using an encryption key?

根据mmkv的flutter版本API设计，设置加密只有通过以下两种方式：
``` dart
// 1
defaultMMKV({String? cryptKey})

// 2
MMKV(String mmapID,
      {MMKVMode mode = MMKVMode.SINGLE_PROCESS_MODE,
      String? cryptKey,
      String? rootDir})
```
我的业务代码是通过第一种创建mmkv对象，在创建对象的时候并没有设置cryptKey；业务代码中也只有这一处调用；
我可以再次上传一份1.2.8版本的mmkv文件。
[mmkv_new.zip](https://github.com/Tencent/MMKV/files/6538168/mmkv_new.zip)

**Comment by lingol** [comment]
>I/MMKV    ( 7406): <MMKV.cpp:156::initialize> version v1.2.7 page size:4096
I/MMKV    ( 7406): <MMKV.cpp:167::initialize> armv8 AES instructions is supported
I/MMKV    ( 7406): <MMKV.cpp:175::initialize> armv8 CRC32 instructions is supported
I/MMKV    ( 7406): <MMKV.cpp:198::initializeMMKV> root dir: /data/user/0/com.tencent.mmkv_example/app_flutter/mmkv
I/flutter ( 7406): MMKV for flutter with rootDir = /data/user/0/com.tencent.mmkv_example/app_flutter/mmkv
I/MMKV    ( 7406): <MMKV_IO.cpp:81::loadFromFile> loading [mmkv.default] with 15834 actual size, file size 16384, InterProcess 0, meta info version:3
I/MMKV    ( 7406): <MMKV_IO.cpp:86::loadFromFile> loading [mmkv.default] with crc 3992902745 sequence 3 version 3
E/MMKV    ( 7406): <MiniPBCoder.cpp:328::decodeOneMap> InvalidProtocolBuffer truncatedMessage
I/MMKV    ( 7406): <MMKV_IO.cpp:130::loadFromFile> loaded [mmkv.default] with 0 key-values
I/flutter ( 7406): all keys = null

And it fails to load, as well. I won't repeat myself unless you at least try something on your side. Like load it in a demo project using MMKV for Flutter v1.2.8.

**Comment by chowjee** [comment]
> > I/MMKV    ( 7406): MMKV.cpp:156::initialize version v1.2.7 page size:4096
> > I/MMKV    ( 7406): MMKV.cpp:167::initialize armv8 AES instructions is supported
> > I/MMKV    ( 7406): MMKV.cpp:175::initialize armv8 CRC32 instructions is supported
> > I/MMKV    ( 7406): MMKV.cpp:198::initializeMMKV root dir: /data/user/0/com.tencent.mmkv_example/app_flutter/mmkv
> > I/flutter ( 7406): MMKV for flutter with rootDir = /data/user/0/com.tencent.mmkv_example/app_flutter/mmkv
> > I/MMKV    ( 7406): <MMKV_IO.cpp:81::loadFromFile> loading [mmkv.default] with 15834 actual size, file size 16384, InterProcess 0, meta info version:3
> > I/MMKV    ( 7406): <MMKV_IO.cpp:86::loadFromFile> loading [mmkv.default] with crc 3992902745 sequence 3 version 3
> > E/MMKV    ( 7406): MiniPBCoder.cpp:328::decodeOneMap InvalidProtocolBuffer truncatedMessage
> > I/MMKV    ( 7406): <MMKV_IO.cpp:130::loadFromFile> loaded [mmkv.default] with 0 key-values
> > I/flutter ( 7406): all keys = null
> 
> And it fails to load, as well. I **won't repeat myself** unless you at least try something on your side. Like loading it in a demo project using MMKV for Flutter v1.2.8.

我提供一个demo

**Comment by chowjee** [comment]
我找到复现问题的路径了，可能和flutter engine版本有关。
我之前使用flutter mmkv 1.2.8版本是基于flutter engine 1.22.5编译的；最近将flutter mmkv升级到1.2.9版本后顺便将flutter engine升级到2.0.5版本，升级后这个问题必现。

**Comment by lingol** [comment]
Did you mean MMKV 1.2.8 with flutter 1.22.5 can load it successfully? Why is that?

**Comment by chowjee** [comment]
> Did you mean MMKV 1.2.8 with flutter 1.22.5 can load it successfully? Why is that?

能加载成功，那是都在同一个flutter版本（1.22.5或者2.0.5）；
同一个flutter版本升级mmkv版本后，mmkv文件是可以正常加载，这个我通过demo验证过了；
但是flutter1.22.5版本，mmkv1.2.8版本写入的数据，在flutter2.0.5版本，mmkv1.2.9版本上读取就失败。

mmkv从1.2.8升级到1.2.9有关于ffi的升级，而flutter从1.22.5升级到2.0.5正好也更新过ffi的API，我怀疑与这个有关系。

提供一个demo，分别在flutter1.22.5/ mmkv 1.2.8写入数据，和在flutter2.0.5/mmkv1.2.9读取数据，就能复现问题。
[demo.zip](https://github.com/Tencent/MMKV/files/6543760/demo.zip)

**Comment by lingol** [comment]
I see. It's tricky.

**Comment by lingol** [comment]
You can get the **default** MMKV on v1.2.9 by passing an **encryption key** `'\u{2}U'`.
```dart
    var mmkv = MMKV.defaultMMKV(cryptKey: '\u{2}U');
```
This bug is due to MMKV's native interface differ from the dart/flutter interface, plus the dart engine's implementation detail differ from 1.2x to 2.x.

**Comment by chowjee** [comment]
> You can get the **default** MMKV on v1.2.9 by passing an **encryption key** `'\u{2}U'`.
> On MMKV for Flutter next version (dev branch):
> 
> ```dart
>     var mmkv = MMKV.defaultMMKV(cryptKey: '\u{2}U');
> ```
> 
> On MMKV for Flutter v1.2.9:
> 
> ```dart
>   var mmkv = MMKV('mmkv.default', cryptKey: '\u{2}U');
> ```
> 
> This bug is due to MMKV's native interface on **Android** differ from the dart/flutter interface, plus the dart engine's implementation detail differ from 1.2x to 2.x.

![image](https://user-images.githubusercontent.com/6858761/119639494-3de1d600-be4a-11eb-9495-f7d2946ad19f.png)
dev分支依赖的com.tencent:mmkv-static:[1.2.9,)一直找不到，请问这个有发布到maven上吗？

**Comment by lingol** [comment]
You can try this instead.
```dart
  var mmkv = MMKV('mmkv.default', cryptKey: '\u{2}U');
```
v1.2.9 will be released shortly.

**Comment by lingol** [comment]
Fixed in MMKV for Flutter v1.2.10, aka MMKV v1.2.9.
